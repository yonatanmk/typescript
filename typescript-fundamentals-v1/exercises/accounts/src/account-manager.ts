interface User {
  email: string;
  password: string;
}

interface ConfirmedUser {
  email: string;
  password: string;
  isActive: boolean;
}

interface Admin {
  email: string;
  password: string;
  isActive: boolean;
  adminSince: Date;
}

export class AccountManager {
  users: User[] = new Array();

  /**
   * Create a new user account
   * @param email email address
   * @param password the user's password
   * @return the new user account. An admin must activate it using activateNewUser
   * @see this.activateNewUser
   */
  register(email: string, password: string): User {
    if(!email) throw 'Must provide an email';
    if(!password) throw 'Must provide a password';
    let user = { email, password };
    this.users.push(user);
    return user;
  }

  /**
   * Activate a new user account
   * @param approver The admin who's approving this new user
   * @param userToApprove Newly-registered user, who is to be activated
   * @return the updated user object, now activated
   */
  activateNewUser(approver: Admin, userToApprove: User): ConfirmedUser {
    if (!approver.adminSince) throw "Approver is not an admin!";
    let toConfirm = userToApprove as ConfirmedUser
    toConfirm.isActive = true;
    return toConfirm;
  }

  /**
   * Promote a normal user to admin
   * @param existingAdmin admin who is promoting another user
   * @param user an active user who you're making an admin
   * @return the updated user object, now can also be regarded as an admin
   */
  promoteToAdmin(existingAdmin: Admin, user: ConfirmedUser): Admin {
    if (!existingAdmin.adminSince) throw "Not an admin!";
    if (user.isActive !== true) throw "User must be active in order to be promoted to admin!";
    let newAdmin = user as Admin
    newAdmin.adminSince = new Date();
    return newAdmin;
  }
}
