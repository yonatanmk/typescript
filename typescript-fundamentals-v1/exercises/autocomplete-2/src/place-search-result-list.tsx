import * as React from 'react';


export const PlaceSearchResultList: React.SFC<any> = function() {
  return (
    <pre>{JSON.stringify(arguments)}</pre>
  )
}
