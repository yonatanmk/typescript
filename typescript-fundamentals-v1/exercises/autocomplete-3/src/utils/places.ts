export interface PlaceSummary {
  description: string;
  id: string;
  place_id: string;
}

export interface PlaceDetails {
  rating: number;
  icon: string;
  name: string;
  url: string;
  vicinity: string;
  website?: string;
}

/**
 * Search the google places API, for a particular term
 *
 * @export
 * @param {string} input Search term
 * @returns {Promise} promise that resolves to search results
 */
export function fetchPlaceSummaries(input: string): Promise<PlaceSummary[]> {
  return fetch(
    `http://localhost:3000/maps/api/place/autocomplete/json?types=establishment&input=${input}`
  )
    .then(response => response.json())
    .then(jsonData => {
      return jsonData.predictions as PlaceSummary[];
    });
}

export function fetchPlaceDetails(placeids: string[]): Promise<PlaceDetails[]> {
  return Promise.all(
    placeids.map(placeid => {
      return fetch(
        `http://localhost:3000/maps/api/place/details/json?placeid=${placeid}`
      )
        .then(response => response.json())
        .then(jsonData => jsonData.result as PlaceDetails);
    })
  );
}