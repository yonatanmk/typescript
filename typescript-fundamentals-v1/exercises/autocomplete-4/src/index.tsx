import * as React from 'react';
import { render } from 'react-dom';
// import { PlaceSearch } from './PlaceSearch';
import { App } from './app';
/**
 * Render the <PlaceSearch> component into the #root element
 */
const root = document.getElementById('root');
render(<App />, root);