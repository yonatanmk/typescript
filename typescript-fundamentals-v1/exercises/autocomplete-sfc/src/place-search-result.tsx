import * as React from 'react';
export const PlaceSearchResult: React.SFC<any> = () => {
  return (
    <div></div>
  );
};