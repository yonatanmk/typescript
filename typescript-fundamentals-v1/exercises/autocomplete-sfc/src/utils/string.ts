
/**
 * Shorten a URL string to a given length 
 * @param originalUrl URL to shorten
 * @param length maximum allowed length
 */
export function shortUrl(originalUrl = '', length = 50) {
  let chunk_l = length / 2;
  let url = originalUrl.replace('http://', '').replace('https://', '');

  if (url.length <= length) {
    return url;
  }

  let start_chunk = shortString(url, chunk_l, false);
  let end_chunk = shortString(url, chunk_l, true);
  return `${start_chunk}..${end_chunk}`;
}

function shortString(str: string, l: number, rev: boolean) {
  let stop_chars = [' ', '/', '&'];
  let acceptable_shortness = l * 0.80; // When to start looking for stop characters
  let reverse = typeof rev !== 'undefined' ? rev : false;
  let s = reverse ? str.split('').reverse().join('') : str;
  let short_s = '';

  for (let i = 0; i < l - 1; i++) {
    short_s += s[i];
    if (i >= acceptable_shortness && stop_chars.indexOf(s[i]) >= 0) {
      break;
    }
  }
  if (reverse) {
    return short_s.split('').reverse().join('');
  }
  return short_s;
}