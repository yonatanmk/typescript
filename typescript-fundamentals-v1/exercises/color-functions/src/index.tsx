import { hexToRgb, rgbToHex } from './color-utils';

export let color = {
  r: 255,
  g: 0,
  b: 0,
  get hex() : string {
    return rgbToHex(this.r, this.g, this.b);
  },
  set hex(hex: string) {
    let { r, g, b } = hexToRgb(hex);
    this.r = r;
    this.g = g;
    this.b = b;
  }
};