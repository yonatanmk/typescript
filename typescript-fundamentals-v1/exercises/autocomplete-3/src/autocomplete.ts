import { wait } from './utils/promise';
import { fetchPlaceSummaries, fetchPlaceDetails, PlaceDetails, PlaceSummary } from './utils/places';

export async function autocomplete(term: string): Promise<PlaceDetails[]> {
  console.log(`⏳ Beginning search for ${term}`);
  // Begin actual query API call
  let placeResults: PlaceSummary[] = await fetchPlaceSummaries(term);
  console.log(`✅ Completed search for ${term}`);
  let placeIds = placeResults.map(place => place.place_id);
  let places: PlaceDetails[] = await fetchPlaceDetails(placeIds);
  console.log(places);
  // Return the results (eventual value of the task)
  return places;
}
