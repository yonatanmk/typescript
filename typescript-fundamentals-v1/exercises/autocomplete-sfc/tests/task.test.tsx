import * as React from 'react';
import { PlaceSearchResult } from '../src/place-search-result';
import * as renderer from 'react-test-renderer';
import { PlaceDetails } from '../src/utils/places';
import { ReactElement } from 'react';

let hasStudentSolution: boolean = true;

hasStudentSolution =
  JSON.stringify(renderer.create(<PlaceSearchResult />).toJSON()) !==
  JSON.stringify({
    nodeType: 'component',
    props: {},
    instance: null,
    rendered: { nodeType: 'host', type: 'div', props: {}, instance: null, rendered: null }
  });

if (hasStudentSolution) {
  it('renders correctly if place is missing a website', () => {
    let pd: PlaceDetails = {
      id: 'e3290ca5b95282e14c0b14f98666ff7886efa359',
      name: 'Donut Hut',
      rating: 1.1,
      icon: 'https://maps.gstatic.com/mapfiles/place_api/icons/restaurant-71.png',
      vicinity: '31 Little Canada Road East, Little Canada',
      url: 'https://maps.google.com/?cid=12061977042674382991'
    };
    const tree = renderer.create(<PlaceSearchResult {...pd} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly if place has a website', () => {
    let pd: PlaceDetails = {
      id: 'e3290ca5b95282e14c0b14f98666ff7886efa359',
      name: 'Donut Hut',
      rating: 1.1,
      icon: 'https://maps.gstatic.com/mapfiles/place_api/icons/restaurant-71.png',
      vicinity: '31 Little Canada Road East, Little Canada',
      url: 'https://maps.google.com/?cid=12061977042674382991',
      website: 'http://example.com'
    };
    const tree = renderer.create(<PlaceSearchResult {...pd} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
} else {
  describe('Instructions', () => {
    test('Please implement the "autocomplete-sfc/src/place-search-result.tsx" component', () => {
      expect(true).toBeTruthy();
    });
  });
}
