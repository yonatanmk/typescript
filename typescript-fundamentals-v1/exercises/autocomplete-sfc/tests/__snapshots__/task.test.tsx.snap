// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly if place has a website 1`] = `
<li
  className="search-result"
>
  <img
    className="icon"
    src="https://maps.gstatic.com/mapfiles/place_api/icons/restaurant-71.png"
  />
  <h3>
     
    Donut Hut
     
  </h3>
  <p>
    <a
      href="https://maps.google.com/?cid=12061977042674382991"
      target="_blank"
    >
      31 Little Canada Road East, Little Canada
    </a>
      -  
    <a
      href="http://example.com"
      target="_blank"
    >
      example.com
    </a>
  </p>
</li>
`;

exports[`renders correctly if place is missing a website 1`] = `
<li
  className="search-result"
>
  <img
    className="icon"
    src="https://maps.gstatic.com/mapfiles/place_api/icons/restaurant-71.png"
  />
  <h3>
     
    Donut Hut
     
  </h3>
  <p>
    <a
      href="https://maps.google.com/?cid=12061977042674382991"
      target="_blank"
    >
      31 Little Canada Road East, Little Canada
    </a>
      -  
    <a
      href={undefined}
      target="_blank"
    >
      
    </a>
  </p>
</li>
`;
